asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py

```diff
@@ -0,0 +1,223 @@
+        allow_mutation=False,
+    )
+
+    @classmethod
+    def from_oemol(
+        cls, mol: oechem.OEMol, compound_name: Optional[str] = None, **kwargs
+    ) -> "Ligand":
+        sdf_str = oemol_to_sdf_string(mol)
+        return cls(data=sdf_str, compound_name=compound_name, **kwargs)
+
+    def to_oemol(self) -> oechem.OEMol:
+        mol = sdf_string_to_oemol(self.data)
+        return mol
+
+    @classmethod
+    def from_smiles(
+        cls, smiles: str, compound_name: Optional[str] = None, **kwargs
+    ) -> "Ligand":
+        mol = smiles_to_oemol(smiles)
+        sdf_str = oemol_to_sdf_string(mol)
+        return cls(data=sdf_str, compound_name=compound_name, **kwargs)
+
+    @property
+    def smiles(self) -> str:
+        mol = sdf_string_to_oemol(self.data)
+        return oemol_to_smiles(mol)
+
+    @property
+    def inchi(self) -> str:
+        mol = sdf_string_to_oemol(self.data)
+        return oemol_to_inchi(mol)
+
+    @property
+    def inchikey(self) -> str:
+        mol = sdf_string_to_oemol(self.data)
+        return oemol_to_inchikey(mol)
+
+    @classmethod
+    def from_sdf(
+        cls,
+        sdf_file: Union[str, Path],
+        compound_name: Optional[str] = None,
+        read_SD_attrs: bool = True,
+        **kwargs,
+    ) -> "Ligand":
+        """
+        Read in a ligand from an SDF file.
+        If read_SD_attrs is True, then SD tags will be read in as attributes, overriding kwargs where double defined.
+        If read_SD_attrs is False, then SD tags will not be read in as attributes, and kwargs will be used instead
+
+        Parameters
+        ----------
+        sdf_file : Union[str, Path]
+            Path to the SDF file
+        compound_name : Optional[str], optional
+            Name of the compound, by default None
+        read_SD_attrs : bool, optional
+            Whether to read in SD tags as attributes, by default True, overrides kwargs
+        """
+        # directly read in data
+        sdf_str = read_file_directly(sdf_file)
+        lig = cls(data=sdf_str, compound_name=compound_name, **kwargs)
+        if read_SD_attrs:
+            lig.pop_attrs_from_SD_data()
+        return lig
+
+    def to_sdf(self, filename: Union[str, Path], write_SD_attrs: bool = True) -> None:
+        """
+        Write out the ligand to an SDF file
+        If write_SD_attrs is True, then SD tags will be written out as attributes
+        If write_SD_attrs is False, then SD tags will not be written out as attributes
+
+        Parameters
+        ----------
+        filename : Union[str, Path]
+            Path to the SDF file
+        write_SD_attrs : bool, optional
+            Whether to write out SD tags as attributes, by default True
+
+        """
+        if write_SD_attrs:
+            self.flush_attrs_to_SD_data()
+        # directly write out data
+        write_file_directly(filename, self.data)
+
+    def set_SD_data(self, data: dict[str, str]) -> None:
```

Comment:
Is there a reason to store these in the sdf string rather than the `Ligand` object under some field like `tags` or `properties`? This would mean that you don't have to convert to openeye mol every time you read and add a tag, you can also still put all of these tags into the molecule when you do convert to oemol as well by extending the `flush_attrs_to_SD_data` method. It just feels a little strange to have a schema to store the info in but the sdf is still the primary storage format. I don't have a feel for what data goes into these SD tags so I could be missing something but could common ones also be added as fields which can be validated?

Reply:
Yeah 10/10 agree with this one! I wonder if we even want to strip the `data` representation of SD Tags to help alleviate the problem you mention below regarding ligands with different SD tags being considered unequal. Then we when we write out the SDF file we can add the SD Tags? idk good thing to argue about on Wednesday :P

Reply:
Basically this comes down to the possibly tricky idea that we are trying to have an SDF file as a first class object here, ie can reconstruct the whole object from one with the correct tags. I think we probably need to make a clear choice one way or another. 

IMO:

**Pros**
-  The chemists work with SDFs not JSON, so if we give them something it can come back with all the info it left with
- No need to carry around other metadata when working with SDFs as will re-serialise to the same object.
- No need to worry about attributes becoming out of sync with underlying sdf str as the underlying data **is** the attributes.  

**Cons**
- A little bit clumsy with custom serialisation logic. 
- Painful roundtripping

**Cons of dropping SDF as serialisable format** 
- Once you write an object to an SDF it will not re-serialize to the same object without also injecting exactly the same metadata. 


If we only want to work with JSON thats fine we can drop attributes as tags all together and things will be much simpler, but the flipside is that once you write out an SDF it is no longer round-trippable. 

I am leaning on the side of keeping SDFs as first class, but we can discuss on Wednesday. 

For reference here is what an `Ligand` full of attributes serializes to as an SDF. 

```

  -OEChem-06282316122D

  6  5  0     0  0  0  0  0  0999 V2000
    0.0000    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.0000    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.5000    0.8660    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.5000    0.8660    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.0000    1.7321    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.0000    1.7321    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  1  0  0  0  0
  3  4  1  0  0  0  0
  4  5  1  0  0  0  0
  5  6  1  0  0  0  0
M  END
> <compound_name>
'MyLig'

> <ids>
{'moonshot_compound_id': '1234', 'manifold_api_id': 'd12508af-afb8-44df-b084-8b536e3f11e2', 'manifold_vc_id': 'BLAH', 'compchem_id': '8a6dc4d2-45ba-48b6-841a-fde29eef3fbc'}

> <experimental_data>
{'compound_id': 'blah2', 'racemic': 'False', 'achiral': 'False', 'absolute_stereochemistry_enantiomerically_pure': 'False', 'relative_stereochemistry_enantiomerically_pure': 'False'}

> <experimental_data_values>
{'pIC50': 1.0}

> <name_jeff>
'42'

$$$$

```

Reply:
Lets discuss in full on Wed.

Reply:
As a partial improvement I have added the `tags` attribute that you can assign to and will be written as SD tags.

Reply:
@JenkeScheen any thoughts? We will probably discuss tommorow.

Reply:
I think we should focus on making the object round-trippable at all times. Let's discuss in more detail on Thursday.